pkg/collector/corechecks/cluster/orchestrator/collector_discovery.go

```diff
@@ -0,0 +1,134 @@
+	"github.com/DataDog/datadog-agent/pkg/orchestrator"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	defaultAPIServerTimeout = 20 * time.Second
+)
+
+// APIServerDiscoveryProvider is a discovery provider that uses the Kubernetes
+// API Server as its data source.
+type APIServerDiscoveryProvider struct {
+	result []collectors.Collector
+	seen   map[string]struct{}
+}
+
+// NewAPIServerDiscoveryProvider returns a new instance of the APIServer
+// discovery provider.
+func NewAPIServerDiscoveryProvider() *APIServerDiscoveryProvider {
+	return &APIServerDiscoveryProvider{
+		seen: make(map[string]struct{}),
+	}
+}
+
+// Discover returns collectors to enable based on information exposed by the API server.
+func (kp *APIServerDiscoveryProvider) Discover(inventory *inventory.CollectorInventory) ([]collectors.Collector, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPIServerTimeout)
+	defer cancel()
+
+	client, err := apiserver.WaitForAPIClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	groups, resources, err := client.DiscoveryCl.ServerGroupsAndResources()
+	if err != nil {
+		if !discovery.IsGroupDiscoveryFailedError(err) {
+			return nil, err
+		}
+		// We don't handle API group errors here because we assume API groups used
+		// by collectors in the orchestrator check will always be part of the result
+		// even though it might be incomplete due to discovery failures on other
+		// groups.
+		for group, apiGroupErr := range err.(*discovery.ErrGroupDiscoveryFailed).Groups {
+			log.Warnf("Resources for API group version %s could not be discovered: %s", group, apiGroupErr)
+		}
+	}
+
+	if len(resources) == 0 {
+		return nil, fmt.Errorf("failed to discover resources from API groups")
+	}
+
+	preferredResources, otherResources := identifyResources(groups, resources)
+
+	// First pass to enable server-preferred resources
+	kp.walkAPIResources(inventory, preferredResources)
+
+	// Second pass to enable other resources
+	kp.walkAPIResources(inventory, otherResources)
+
+	return kp.result, nil
+}
+
+func (kp *APIServerDiscoveryProvider) addCollector(collector collectors.Collector) {
+	// Make sure resource collectors are added at most once
+	if _, found := kp.seen[collector.Metadata().Name]; found {
+		return
+	}
+
+	kp.result = append(kp.result, collector)
+	kp.seen[collector.Metadata().Name] = struct{}{}
+	log.Debugf("Discovered collector %s", collector.Metadata().FullName())
+}
+
+func (kp *APIServerDiscoveryProvider) walkAPIResources(inventory *inventory.CollectorInventory, resources []*v1.APIResourceList) {
+	for _, list := range resources {
+		for _, resource := range list.APIResources {
+			collector, err := inventory.CollectorForVersion(resource.Name, list.GroupVersion)
+			if err != nil {
+				continue
+			}
+
+			// Enable the cluster collector when the node resource is discovered.
```

Comment:
shouldn't we support that the other way around as well?

Reply:
there's no cluster resource in kubernetes

Reply:
What I mean is what if someone sets this up in their config without the nodes collector:
```
  - collectors:
    - clusters
```

Shouldn't we automatically collect nodes as well?

Reply:
in that case we don't use collector discovery but just the check config since it has precedence
so we would be running the same code as of today, which expects `cluster` to be part of the list

I didn't plan to update this since it's not really in the scope of that PR
I think in case we want to update it, that would make more sense to do it in a new PR than included here (for visibility at least)

Reply:
Yes makes sense to me! Just wanted to check whether I got the idea correct

Reply:
sure 👍 

my personal preference goes to explicit declaration of the cluster resource since it's a more predictable configuration for customers and it's easier to maintain for us